src/v/cloud_roles/aws_sts_refresh_impl.cc

```diff
@@ -0,0 +1,166 @@
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/redpanda-data/redpanda/blob/master/licenses/rcl.md
+ */
+
+#include "cloud_roles/aws_sts_refresh_impl.h"
+
+#include "s3/client.h"
+#include "utils/file_io.h"
+
+#include <boost/property_tree/ptree.hpp>
+
+/// The path to credentials in the XML response returned by STS
+static constexpr std::string_view response_node_credential_path
+  = "AssumeRoleWithWebIdentityResponse.AssumeRoleWithWebIdentityResult."
+    "Credentials";
+
+/// These variables injected by the AWS operator must be present on the redpanda
+/// pod to ensure we can get credentials from STS.
+/// ref:
+/// https://docs.amazonaws.cn/en_us/eks/latest/userguide/specify-service-account-role.html
+struct aws_injected_env_vars {
+    static constexpr std::string_view role_arn = "AWS_ROLE_ARN";
+
+    static constexpr std::string_view token_file_path
+      = "AWS_WEB_IDENTITY_TOKEN_FILE";
+};
+
+struct sts_params {
+    static constexpr std::string_view role_session_name = "redpanda_si";
+
+    static constexpr std::string_view api_version = "2011-06-15";
+
+    static constexpr std::chrono::seconds token_expiry_seconds{3599};
+};
+
+struct sts_response_schema {
+    static constexpr std::string_view expiry = "Expiration";
+    static constexpr std::string_view access_key_id = "AccessKeyId";
+    static constexpr std::string_view secret_access_key = "SecretAccessKey";
+    static constexpr std::string_view session_token = "SessionToken";
+};
+
+static const boost::beast::string_view content_type{
+  "application/x-www-form-urlencoded"};
+
+static constexpr std::string_view request_payload
+  = "Action=AssumeRoleWithWebIdentity"
+    "&DurationSeconds={}"
+    "&Version={}"
+    "&RoleSessionName={}"
+    "&RoleArn={}"
+    "&WebIdentityToken={}";
+
+static ss::sstring load_from_env(std::string_view env_var) {
+    auto env_value = std::getenv(env_var.data());
+    if (!env_value) {
+        throw std::runtime_error(fmt::format(
+          "environment variable {} is not set, the STS client cannot function "
+          "without this.",
+          env_var));
+    }
+    return env_value;
+}
+
+cloud_roles::aws_sts_refresh_impl::aws_sts_refresh_impl(
+  seastar::sstring api_host,
+  uint16_t api_port,
+  s3::aws_region_name region,
+  ss::abort_source& as,
+  retry_params retry_params)
+  : refresh_credentials::impl(
+    std::move(api_host), api_port, std::move(region), as, retry_params)
+  , _role{load_from_env(aws_injected_env_vars::role_arn)}
+  , _token_file_path{load_from_env(aws_injected_env_vars::token_file_path)} {}
+
+ss::future<cloud_roles::api_response>
+cloud_roles::aws_sts_refresh_impl::fetch_credentials() {
+    // The content of the token file is periodically rotated by k8s, so we
+    // read it fully each time to avoid stale tokens
+
+    ss::sstring token{};
+    try {
+        token = co_await read_fully_to_string({_token_file_path});
+    } catch (...) {
```

Comment:
Maybe we can be more specific then this, and not just `throw`

Reply:
Yes good point, I plan to add a new kind of error to be returned here instead of throwing, for infra related issues.